### PR TITLE
Improve substitution performance by fixing type instabilities

### PR DESCRIPTION
**Pull request by rdeits**
This yields about a 10x improvement in substitution time on my simple benchmark (included in perf/runbenchmarks.jl).

### CODE REVIEW

File: src/subs.jl [hunk 1]
```diff
@@ -1,5 +1,5 @@
-function evalmap(vars, x::Vector, varorder::Vector{PolyVar})
-  vals = Any[var for var in vars]
+function evalmap{T}(vars, x::Vector{T}, varorder::Vector{PolyVar})
```
Comment (blegat) [inline src/subs.jl]: Should we do a shortcut like
```
if vars == varorder
x
else
... # current body of evalmap
end
```

Comment (rdeits) [inline src/subs.jl]: good idea; done

**Comment by rdeits** [comment]
Okay, this has been rebased and cleaned up.

**Comment by blegat** [comment]
Thanks a lot for looking into this !

**Comment by rdeits** [comment]
No problem :). Optimizing julia code is surprisingly satisfying.

**Pull request merged**